### PR TITLE
Fix data races

### LINKED ISSUE #113: Race found with Node.PMax
Hi hashis,

Found this data race while using `memberlist` as of commit 23ad4b7d7b38496cd64c241dfd4c60b7794c254a :
```==================
WARNING: DATA RACE
Read at 0x00c4211acf49 by goroutine 660:
  github.com/fastly/checkup/vendor/github.com/hashicorp/memberlist.(*Memberlist).rawSendMsgUDP()
      /home/vagrant/devel/go-workspace/src/github.com/fastly/checkup/vendor/github.com/hashicorp/memberlist/net.go:662 +0x63b
  github.com/fastly/checkup/vendor/github.com/hashicorp/memberlist.(*Memberlist).gossip()
      /home/vagrant/devel/go-workspace/src/github.com/fastly/checkup/vendor/github.com/hashicorp/memberlist/state.go:509 +0x733
  github.com/fastly/checkup/vendor/github.com/hashicorp/memberlist.(*Memberlist).(github.com/fastly/checkup/vendor/github.com/hashicorp/memberlist.gossip)-fm()
      /home/vagrant/devel/go-workspace/src/github.com/fastly/checkup/vendor/github.com/hashicorp/memberlist/state.go:94 +0x41
  github.com/fastly/checkup/vendor/github.com/hashicorp/memberlist.(*Memberlist).triggerFunc()
      /home/vagrant/devel/go-workspace/src/github.com/fastly/checkup/vendor/github.com/hashicorp/memberlist/state.go:118 +0x159

Previous write at 0x00c4211acf49 by goroutine 705:
  github.com/fastly/checkup/vendor/github.com/hashicorp/memberlist.(*Memberlist).aliveNode()
      /home/vagrant/devel/go-workspace/src/github.com/fastly/checkup/vendor/github.com/hashicorp/memberlist/state.go:952 +0xef2
  github.com/fastly/checkup/vendor/github.com/hashicorp/memberlist.(*Memberlist).handleAlive()
      /home/vagrant/devel/go-workspace/src/github.com/fastly/checkup/vendor/github.com/hashicorp/memberlist/net.go:558 +0x32f
  github.com/fastly/checkup/vendor/github.com/hashicorp/memberlist.(*Memberlist).udpHandler()
      /home/vagrant/devel/go-workspace/src/github.com/fastly/checkup/vendor/github.com/hashicorp/memberlist/net.go:410 +0x308

Goroutine 660 (running) created at:
  github.com/fastly/checkup/vendor/github.com/hashicorp/memberlist.(*Memberlist).schedule()
      /home/vagrant/devel/go-workspace/src/github.com/fastly/checkup/vendor/github.com/hashicorp/memberlist/state.go:94 +0x423
  github.com/fastly/checkup/vendor/github.com/hashicorp/memberlist.Create()
      /home/vagrant/devel/go-workspace/src/github.com/fastly/checkup/vendor/github.com/hashicorp/memberlist/memberlist.go:164 +0xcd
  github.com/fastly/checkup/vendor/github.com/hashicorp/serf/serf.Create()
      /home/vagrant/devel/go-workspace/src/github.com/fastly/checkup/vendor/github.com/hashicorp/serf/serf/serf.go:398 +0x13d5
  github.com/fastly/checkup/cluster.New()
      /home/vagrant/devel/go-workspace/src/github.com/fastly/checkup/cluster/cluster.go:126 +0x6e0
  github.com/fastly/checkup/checkup.New()
      /home/vagrant/devel/go-workspace/src/github.com/fastly/checkup/checkup/checkup.go:206 +0x13a7
  github.com/fastly/checkup/test.StartCheckup()
      /home/vagrant/devel/go-workspace/src/github.com/fastly/checkup/test/helper.go:62 +0x36f
  command-line-arguments_test.startManyCheckupsAndFakeVarnishes()
      /home/vagrant/devel/go-workspace/checkup/fakevarnish/scale_test.go:333 +0x534
  command-line-arguments_test.TestManyCheckupsWithDifferentBackends()
      /home/vagrant/devel/go-workspace/checkup/fakevarnish/scale_test.go:166 +0x20f
  testing.tRunner()
      /home/vagrant/devel/go-workspace/go/src/testing/testing.go:657 +0x107

Goroutine 705 (running) created at:
  github.com/fastly/checkup/vendor/github.com/hashicorp/memberlist.newMemberlist()
      /home/vagrant/devel/go-workspace/src/github.com/fastly/checkup/vendor/github.com/hashicorp/memberlist/memberlist.go:146 +0xde0
  github.com/fastly/checkup/vendor/github.com/hashicorp/memberlist.Create()
      /home/vagrant/devel/go-workspace/src/github.com/fastly/checkup/vendor/github.com/hashicorp/memberlist/memberlist.go:156 +0x3c
  github.com/fastly/checkup/vendor/github.com/hashicorp/serf/serf.Create()
      /home/vagrant/devel/go-workspace/src/github.com/fastly/checkup/vendor/github.com/hashicorp/serf/serf/serf.go:398 +0x13d5
  github.com/fastly/checkup/cluster.New()
      /home/vagrant/devel/go-workspace/src/github.com/fastly/checkup/cluster/cluster.go:126 +0x6e0
  github.com/fastly/checkup/checkup.New()
      /home/vagrant/devel/go-workspace/src/github.com/fastly/checkup/checkup/checkup.go:206 +0x13a7
  github.com/fastly/checkup/test.StartCheckup()
      /home/vagrant/devel/go-workspace/src/github.com/fastly/checkup/test/helper.go:62 +0x36f
  command-line-arguments_test.startManyCheckupsAndFakeVarnishes()
      /home/vagrant/devel/go-workspace/checkup/fakevarnish/scale_test.go:333 +0x534
  command-line-arguments_test.TestManyCheckupsWithDifferentBackends()
      /home/vagrant/devel/go-workspace/checkup/fakevarnish/scale_test.go:166 +0x20f
  testing.tRunner()
      /home/vagrant/devel/go-workspace/go/src/testing/testing.go:657 +0x107
==================
```

I'm not sure if `nodeLock` (https://github.com/hashicorp/memberlist/blob/902b55b5d29084b62aa0d7989ed7f89ee7e9230e/memberlist.go#L47) is intended to protect just `nodeMap` in `aliveNode()` or also the `nodeState ` / `Node`s from the map and their data (such as `PMax`).

Looks like potential races also in `probeNode`:
https://github.com/hashicorp/memberlist/blob/master/state.go#L326
https://github.com/hashicorp/memberlist/blob/master/state.go#L347

### LINKED ISSUE #134: Data Races
@slackpad a data race on `Incarnation`. Maybe you have an idea on how to address this:

```
WARNING: DATA RACE
Write at 0x00c424b902d0 by goroutine 5112:
  github.com/hashicorp/consul/vendor/github.com/hashicorp/memberlist.(*Memberlist).refute()
      /Users/frank/go/consul/src/github.com/hashicorp/consul/vendor/github.com/hashicorp/memberlist/state.go:807 +0x7f
  github.com/hashicorp/consul/vendor/github.com/hashicorp/memberlist.(*Memberlist).suspectNode()
      /Users/frank/go/consul/src/github.com/hashicorp/consul/vendor/github.com/hashicorp/memberlist/state.go:1033 +0xa41
  github.com/hashicorp/consul/vendor/github.com/hashicorp/memberlist.(*Memberlist).handleSuspect()
      /Users/frank/go/consul/src/github.com/hashicorp/consul/vendor/github.com/hashicorp/memberlist/net.go:518 +0x297
  github.com/hashicorp/consul/vendor/github.com/hashicorp/memberlist.(*Memberlist).packetHandler()
      /Users/frank/go/consul/src/github.com/hashicorp/consul/vendor/github.com/hashicorp/memberlist/net.go:384 +0x360

Previous read at 0x00c424b902d0 by goroutine 168:
  github.com/hashicorp/consul/vendor/github.com/hashicorp/memberlist.(*Memberlist).Leave()
      /Users/frank/go/consul/src/github.com/hashicorp/consul/vendor/github.com/hashicorp/memberlist/memberlist.go:580 +0x269
  github.com/hashicorp/consul/vendor/github.com/hashicorp/serf/serf.(*Serf).Leave()
      /Users/frank/go/consul/src/github.com/hashicorp/consul/vendor/github.com/hashicorp/serf/serf/serf.go:690 +0x2fc
  github.com/hashicorp/consul/agent/consul.(*Client).Leave()
      /Users/frank/go/consul/src/github.com/hashicorp/consul/agent/consul/client.go:180 +0xf9
  github.com/hashicorp/consul/agent/consul.TestLeader_LeftMember()
      /Users/frank/go/consul/src/github.com/hashicorp/consul/agent/consul/leader_test.go:178 +0x25b
  testing.tRunner()
      /Users/frank/go1.9/src/testing/testing.go:746 +0x16c

Goroutine 5112 (running) created at:
  github.com/hashicorp/consul/vendor/github.com/hashicorp/memberlist.newMemberlist()
      /Users/frank/go/consul/src/github.com/hashicorp/consul/vendor/github.com/hashicorp/memberlist/memberlist.go:180 +0x895
  github.com/hashicorp/consul/vendor/github.com/hashicorp/memberlist.Create()
      /Users/frank/go/consul/src/github.com/hashicorp/consul/vendor/github.com/hashicorp/memberlist/memberlist.go:190 +0x3c
  github.com/hashicorp/consul/vendor/github.com/hashicorp/serf/serf.Create()
      /Users/frank/go/consul/src/github.com/hashicorp/consul/vendor/github.com/hashicorp/serf/serf/serf.go:395 +0x13b7
  github.com/hashicorp/consul/agent/consul.(*Client).setupSerf()
      /Users/frank/go/consul/src/github.com/hashicorp/consul/agent/consul/client_serf.go:47 +0xe8b
  github.com/hashicorp/consul/agent/consul.NewClientLogger()
      /Users/frank/go/consul/src/github.com/hashicorp/consul/agent/consul/client.go:138 +0xa80
  github.com/hashicorp/consul/agent/consul.NewClient()
      /Users/frank/go/consul/src/github.com/hashicorp/consul/agent/consul/client.go:80 +0x41
  github.com/hashicorp/consul/agent/consul.testClientWithConfig()
      /Users/frank/go/consul/src/github.com/hashicorp/consul/agent/consul/client_test.go:57 +0x78
  github.com/hashicorp/consul/agent/consul.testClient()
      /Users/frank/go/consul/src/github.com/hashicorp/consul/agent/consul/client_test.go:39 +0x85
  github.com/hashicorp/consul/agent/consul.TestLeader_LeftMember()
      /Users/frank/go/consul/src/github.com/hashicorp/consul/agent/consul/leader_test.go:157 +0xc7
```

### PR DESCRIPTION
**Pull request by dnephin**
Fixes #134
Fixes #113
Closes #179

This PR fixes a number of data races and runs tests in CI using `go test -race` to catch any regressions. Also updates the version of Go used in CI to 1.15. I tried upgrading to 1.16 but there are a number of problems (for some reason the $PATH in that 1.16 image is wrong, and `go vet` reports more problems https://golang.org/doc/go1.16#vet, so I'll defer that upgrade to another PR).

Lots of detail in the commit messages, best viewed by individual commit.

The `EventDelegate` is really only a partial fix for these tests. The data  race is still there, and unfortunately it seems like the interface is a problem.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -13,20 +13,8 @@ jobs:
     executor: go
     steps:
       - checkout
-
-      # Restore go module cache if there is one
-      - restore_cache:
-          keys:
-            - memberlist-modcache-v1-{{ checksum "go.mod" }}
```
Comment (dnephin) [inline .circleci/config.yml:20]: With the introduction of go modules it is actually faster to download just the module from the go proxy than to try and cache the local `/go/pkg/mod`.

File: event_delegate.go [hunk 2]
```diff
@@ -49,13 +49,16 @@ type NodeEvent struct {
 }
 
 func (c *ChannelEventDelegate) NotifyJoin(n *Node) {
-	c.Ch <- NodeEvent{NodeJoin, n}
+	node := *n
```
Comment (rboyer) [inline event_delegate.go:52]: Note: This is a shallow copy of `Node.Meta` which is a `[]byte`

File: .circleci/config.yml [hunk 3]
```diff
@@ -52,8 +40,7 @@ jobs:
 
       # run go tests with gotestsum
       - run: |
-          PACKAGE_NAMES=$(go list ./...)
-          gotestsum --format=short-verbose --junitfile $TEST_RESULTS/gotestsum-report.xml -- $PACKAGE_NAMES
+          gotestsum --format=short-verbose --junitfile $TEST_RESULTS/gotestsum-report.xml -- -race
```
Comment (dnephin) [inline .circleci/config.yml:43]: The default list of packages for `gotestsum` is `./...`, so I removed that arg.

File: event_delegate.go (hunk 2 above)
Comment (dnephin) [reply event_delegate.go:52]: Ah yes, good point! This interface is still racy, so this is only a quick hack to get the tests passing the race detector. We will need to make a larger breaking change to this interface to fix it properly for serf and Consul.

File: memberlist_test.go [hunk 4]
```diff
@@ -1494,7 +1485,6 @@ func TestMemberlist_Join_IPv6(t *testing.T) {
 	c2.Name = "B"
 	c2.BindAddr = "[::1]"
 	c2.BindPort = 0 // choose free
-	c2.Logger = testLoggerWithName(t, c2.Name)
```
[the remark this answers is not in the document]
Comment (dnephin) [reply memberlist_test.go:1497]: Aha, good call out. I will restore the naming part of the logger, but continue to use stdout/stderr as the writer.

[the remark this answers is not in the document]
Comment (dnephin) [reply memberlist_test.go:1497]: Ok, I've restored the name to most of the loggers. There was one test that only creates a single instance, so I left that one with the default logger.

File: event_delegate.go (hunk 2 above)
Comment (rboyer) [reply event_delegate.go:52]: :+1:

**Review by rboyer — APPROVED** [review]
LGTM

**Pull request merged**